packages/devtools_app/lib/src/flutter/banner_messages.dart

```diff
@@ -0,0 +1,304 @@
+
+const _profileGranularityDocsUrl =
+    'https://flutter.dev/docs/development/tools/devtools/performance#profile-granularity';
+
+class BannerMessagesController implements DisposableController {
+  final _dismissedMessageIds = <String>{};
+
+  final _refreshMessagesController = StreamController<Null>.broadcast();
+
+  Stream<Null> get onRefreshMessages => _refreshMessagesController.stream;
+
+  @override
+  void dispose() {
+    _refreshMessagesController.close();
+  }
+
+  void refreshMessages() {
+    _refreshMessagesController.add(null);
+  }
+
+  void dismissMessage(String messageId) {
+    assert(!isMessageDismissed(messageId));
+    _dismissedMessageIds.add(messageId);
+    refreshMessages();
+  }
+
+  bool isMessageDismissed(String messageId) {
+    return _dismissedMessageIds.contains(messageId);
+  }
+}
+
+class BannerMessageContainer extends StatefulWidget {
+  const BannerMessageContainer({Key key, @required this.screen})
+      : super(key: key);
+
+  final Screen screen;
+
+  @override
+  _BannerMessageContainerState createState() => _BannerMessageContainerState();
+}
+
+class _BannerMessageContainerState extends State<BannerMessageContainer>
+    with AutoDisposeMixin {
+  BannerMessagesController controller;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    final newController = Controllers.of(context)?.bannerMessages;
+    if (newController == controller) return;
+    controller = newController;
+
+    autoDispose(controller.onRefreshMessages.listen((_) => setState(() {})));
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final messagesForScreen = widget.screen.messages(context);
+    if (messagesForScreen.isNotEmpty) {
+      final messagesToShow = <Widget>[];
+      for (Widget message in messagesForScreen) {
+        assert(message is UniqueMessage);
+        if (!controller.isMessageDismissed((message as UniqueMessage).id)) {
+          messagesToShow.add(message);
+        }
+      }
+      if (messagesToShow.isNotEmpty) {
+        return Column(
+          children: messagesToShow,
+        );
+      }
+    }
+    return const SizedBox();
+  }
+}
+
+class BannerMessage extends StatelessWidget implements UniqueMessage {
+  const BannerMessage({
+    @required this.messageId,
+    @required this.textSpans,
+    @required this.backgroundColor,
+    @required this.foregroundColor,
+  });
+
+  final String messageId;
+  final List<TextSpan> textSpans;
+  final Color backgroundColor;
+  final Color foregroundColor;
+
+  @override
+  String get id => messageId;
+
+  @override
+  Widget build(BuildContext context) {
+    final bannerMessagesController = Controllers.of(context).bannerMessages;
+    return Card(
+      color: backgroundColor,
+      child: Padding(
+        padding: const EdgeInsets.all(defaultSpacing),
+        child: Row(
+          children: <Widget>[
+            Expanded(
+              child: Container(
```

Comment:
@pq a lint should be able to catch that this Container widget can be removed.

Reply:
Indeed! [`avoid_unnecessary_containers`](https://dart-lang.github.io/linter/lints/avoid_unnecessary_containers.html) should do just the trick.

Reply:
(PR enabling it on the way...)

Reply:
https://github.com/flutter/devtools/pull/1802